server/server.go

```diff
@@ -851,7 +863,130 @@ func (s *Server) handleExit(ctx context.Context, event fsnotify.Event) {
 	if err := s.ContainerStateToDisk(ctx, c); err != nil {
 		log.Warnf(ctx, "Unable to write %s %s state to disk: %v", resource, c.ID(), err)
 	}
+	s.generateCRIEvent(ctx, c, types.ContainerEventType_CONTAINER_STOPPED_EVENT)
 	if err := os.Remove(event.Name); err != nil {
 		log.Warnf(ctx, "Failed to remove exit file: %v", err)
 	}
 }
+
+func (s *Server) getSandboxStatuses(ctx context.Context, sandboxId string) (*types.PodSandboxStatus, error) {
+	sandboxStatusRequest := &types.PodSandboxStatusRequest{PodSandboxId: sandboxId}
+	sandboxStatus, err := s.PodSandboxStatus(ctx, sandboxStatusRequest)
+
+	if isNotFound(err) {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("getSandboxStatuses: %w", err)
+	}
+
+	return sandboxStatus.GetStatus(), nil
+}
+
+func (s *Server) getContainerStatuses(ctx context.Context, sandboxUid string) ([]*types.ContainerStatus, error) {
+	listContainerRequest := &types.ListContainersRequest{Filter: &types.ContainerFilter{LabelSelector: map[string]string{"io.kubernetes.pod.uid": sandboxUid}}}
+	containers, err := s.ListContainers(ctx, listContainerRequest)
+
+	if err != nil {
+		return []*types.ContainerStatus{}, err
+	}
+
+	containerStatuses := []*types.ContainerStatus{}
+	for _, cc := range containers.GetContainers() {
+		containerStatusRequest := &types.ContainerStatusRequest{ContainerId: cc.Id}
+		resp, err := s.ContainerStatus(ctx, containerStatusRequest)
+		if isNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return []*types.ContainerStatus{}, err
+		}
+		containerStatuses = append(containerStatuses, resp.GetStatus())
+
+	}
+
+	return containerStatuses, nil
+}
+
+func (s *Server) getContainerStatusesFromSandboxId(ctx context.Context, sandboxId string) ([]*types.ContainerStatus, error) {
+	listContainerRequest := &types.ListContainersRequest{Filter: &types.ContainerFilter{PodSandboxId: sandboxId}}
+	containers, err := s.ListContainers(ctx, listContainerRequest)
+
+	if err != nil {
+		return []*types.ContainerStatus{}, err
+	}
+
+	containerStatuses := []*types.ContainerStatus{}
+	for _, cc := range containers.GetContainers() {
+		containerStatusRequest := &types.ContainerStatusRequest{ContainerId: cc.Id, Verbose: false}
+		resp, err := s.ContainerStatus(ctx, containerStatusRequest)
+		if isNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return []*types.ContainerStatus{}, err
+		}
+		containerStatuses = append(containerStatuses, resp.GetStatus())
+
+	}
+
+	return containerStatuses, nil
+}
+
+func (s *Server) generateCRIEvent(ctx context.Context, container *oci.Container, eventType types.ContainerEventType) error {
```

Comment:
No need to return the error here if we do not check it anywhere. :)

Reply:
I agree, but in future, if we have to rely on the returned error (Ex: metric calculation?) in the respective functions, I thought this signature would help. WDYT?

Reply:
Hm, the linter will definitely complain that the error is not checked.

Reply:
Addressed